In [1]:
import pandas as pd

In [2]:
train = pd.read_csv('6722-credit-card-default-data.csv')
test = pd.read_csv('6722-pilot-customers.csv')

In [3]:
train.shape, test.shape

((24000, 25), (1000, 24))

In [4]:
train.isnull().sum().sum(), train.isnull().sum().sum()

(0, 0)

In [5]:
train.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_1  PAY_2  PAY_3  PAY_4  \
0   1    20000.0    2          2         1   24      2      2     -1     -1   
1   2    90000.0    2          2         2   34      0      0      0      0   
2   3    50000.0    2          2         1   37      0      0      0      0   
3   4    50000.0    1          2         1   57     -1      0     -1      0   
4   5    50000.0    1          1         2   37      0      0      0      0   

   ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  ...        0.0        0.0        0.0       0.0     689.0       0.0   
1  ...    14331.0    14948.0    15549.0    1518.0    1500.0    1000.0   
2  ...    28314.0    28959.0    29547.0    2000.0    2019.0    1200.0   
3  ...    20940.0    19146.0    19131.0    2000.0   36681.0   10000.0   
4  ...    19394.0    19619.0    20024.0    2500.0    1815.0     657.0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  DEFAULT  
0       0.0       0.0       0.0        1  
1    1000.0    1000.0    5000.0        0  
2    1100.0    1069.0    1000.0        0  
3    9000.0     689.0     679.0        0  
4    1000.0    1000.0     800.0        0  

[5 rows x 25 columns]

In [6]:
test.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_1  PAY_2  PAY_3  \
0  n1000-1   500000.0    1          1         2   29      0      0      0   
1  n1000-2   210000.0    1          1         2   29     -2     -2     -2   
2  n1000-3   150000.0    1          1         2   27      0      0      0   
3  n1000-4    20000.0    1          2         1   38      0      0      0   
4  n1000-5   400000.0    1          2         1   34     -1     -1     -1   

   PAY_4  ...  BILL_AMT3  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0      0  ...     445007     542653     483003     473944   55000.0   40000.0   
1     -2  ...          0          0          0          0       0.0       0.0   
2      0  ...      89006      89775      87725      40788    4031.0   10006.0   
3      0  ...      19559      18240      17928        150    1699.0    1460.0   
4     -1  ...      11867       7839      14837       7959    9677.0   11867.0   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  
0   38000.0     20239   13750.0     13770  
1       0.0         0       0.0         0  
2    3266.0      4040    1698.0       800  
3     626.0      1750     150.0         0  
4    7839.0     14837    7959.0      5712  

[5 rows x 24 columns]

# Code using ChatGPT

## Prompt:

I have two pandas dataframe in train and test variable:

train:
	ID	LIMIT_BAL	SEX	EDUCATION	MARRIAGE	AGE	PAY_1	PAY_2	PAY_3	PAY_4	...	BILL_AMT4	BILL_AMT5	BILL_AMT6	PAY_AMT1	PAY_AMT2	PAY_AMT3	PAY_AMT4	PAY_AMT5	PAY_AMT6	DEFAULT
0	1	20000.0	2	2	1	24	2	2	-1	-1	...	0.0	0.0	0.0	0.0	689.0	0.0	0.0	0.0	0.0	1
1	2	90000.0	2	2	2	34	0	0	0	0	...	14331.0	14948.0	15549.0	1518.0	1500.0	1000.0	1000.0	1000.0	5000.0	0
2	3	50000.0	2	2	1	37	0	0	0	0	...	28314.0	28959.0	29547.0	2000.0	2019.0	1200.0	1100.0	1069.0	1000.0	0
3	4	50000.0	1	2	1	57	-1	0	-1	0	...	20940.0	19146.0	19131.0	2000.0	36681.0	10000.0	9000.0	689.0	679.0	0
4	5	50000.0	1	1	2	37	0	0	0	0	...	19394.0	19619.0	20024.0	2500.0	1815.0	657.0	1000.0	1000.0	800.0	0
5 rows × 25 columns

test:
	ID	LIMIT_BAL	SEX	EDUCATION	MARRIAGE	AGE	PAY_1	PAY_2	PAY_3	PAY_4	...	BILL_AMT3	BILL_AMT4	BILL_AMT5	BILL_AMT6	PAY_AMT1	PAY_AMT2	PAY_AMT3	PAY_AMT4	PAY_AMT5	PAY_AMT6
0	n1000-1	500000.0	1	1	2	29	0	0	0	0	...	445007	542653	483003	473944	55000.0	40000.0	38000.0	20239	13750.0	13770
1	n1000-2	210000.0	1	1	2	29	-2	-2	-2	-2	...	0	0	0	0	0.0	0.0	0.0	0	0.0	0
2	n1000-3	150000.0	1	1	2	27	0	0	0	0	...	89006	89775	87725	40788	4031.0	10006.0	3266.0	4040	1698.0	800
3	n1000-4	20000.0	1	2	1	38	0	0	0	0	...	19559	18240	17928	150	1699.0	1460.0	626.0	1750	150.0	0
4	n1000-5	400000.0	1	2	1	34	-1	-1	-1	-1	...	11867	7839	14837	7959	9677.0	11867.0	7839.0	14837	7959.0	5712
5 rows × 24 columns

My goal is to build a classification model to predict DEFAULT column. Training must happen on train data and after that test data should be used for prediction. At the end the output I want must contain two columns:

ID DEFAULT
n1000-1 1
n1000-2 0
....

The model must be optimized and selected based on AUC. Use sklearn to write the code. You must be careful about following things:

- categorical encoding
- numerical data preprocessing as needed
- any feature engineering
- anything else that can improve the performance of model

In [7]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score, make_scorer

# Define features and target
X_train = train.drop(columns=['ID', 'DEFAULT'])
y_train = train['DEFAULT']
X_test = test.drop(columns=['ID'])

# Preprocess categorical and numerical features
categorical_features = ['SEX', 'EDUCATION', 'MARRIAGE']
numerical_features = X_train.select_dtypes(include=['float64', 'int64']).columns.difference(categorical_features)

# Define preprocessing steps
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

# Combine preprocessing steps
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_features),
        ('cat', categorical_transformer, categorical_features)
    ])

# Define the model
model = RandomForestClassifier(random_state=42)

# Create a pipeline that preprocesses the data and then fits the model
clf = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', model)
])

# Define the parameter grid for GridSearchCV
param_grid = {
    'classifier__n_estimators': [50, 100, 200],
    'classifier__max_depth': [None, 10, 20, 30],
    'classifier__min_samples_split': [2, 5, 10]
}

# Use AUC as the scoring metric
scorer = make_scorer(roc_auc_score, greater_is_better=True)

# Initialize GridSearchCV
grid_search = GridSearchCV(clf, param_grid, cv=5, scoring=scorer, n_jobs=-1)

# Fit the model
grid_search.fit(X_train, y_train)

# Get the best model
best_model = grid_search.best_estimator_

# Make predictions on the test set
test_predictions = best_model.predict(X_test)

# Prepare the output dataframe
output = pd.DataFrame({
    'ID': test['ID'],
    'DEFAULT': test_predictions
})

# Save the output to a CSV file
output.to_csv('predictions.csv', index=False)
print(output.head())


        ID  DEFAULT
0  n1000-1        0
1  n1000-2        0
2  n1000-3        0
3  n1000-4        0
4  n1000-5        0
